Java 11 build failure was unrelated:
```
kafka.api.UserQuotaTest.testThrottledProducerConsumer


java.lang.AssertionError: Client with id=QuotasTestProducer-1 should have been throttled
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at kafka.api.QuotaTestClients.verifyThrottleTimeMetric(BaseQuotaTest.scala:229)
	at kafka.api.QuotaTestClients.verifyProduceThrottle(BaseQuotaTest.scala:215)
	at kafka.api.BaseQuotaTest.testThrottledProducerConsumer(BaseQuotaTest.scala:82)
```

Java 8 failure was related: we can't use SAM conversion in this branch. I'll roll that part of the change back.

--- a/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala
+++ b/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala
@@ -35,6 +35,7 @@ import org.junit.experimental.categories.Category
  * Note: In the current project settings SAM type conversion is turned off as it's experimental in Scala 2.11.
  * Hence the native Java API based version is more verbose.
  */
+//noinspection ConvertExpressionToSAM
 @Category(Array(classOf[IntegrationTest]))
 class StreamToTableJoinScalaIntegrationTestImplicitSerdes extends StreamToTableJoinScalaIntegrationTestBase {
 
@@ -169,7 +170,7 @@ class StreamToTableJoinScalaIntegrationTestImplicitSerdes extends StreamToTableJ
       .groupByKey(Grouped.`with`[String, JLong](Serdes.String, Serdes.JavaLong))
       .reduce {
         new Reducer[JLong] {
-          def apply(v1: JLong, v2: JLong) = v1 + v2
+          def apply(v1: JLong, v2: JLong): JLong = v1 + v2
         }
       }
 

--- a/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala
+++ b/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala
@@ -157,15 +157,21 @@ class StreamToTableJoinScalaIntegrationTestImplicitSerdes extends StreamToTableJ
 
     // Change the stream from <user> -> <region, clicks> to <region> -> <clicks>
     val clicksByRegion: KStreamJ[String, JLong] = userClicksJoinRegion
-      .map(
-        (_: String, regionWithClicks: (String, JLong)) =>
-          new KeyValue[String, JLong](regionWithClicks._1, regionWithClicks._2)
-      )
+      .map {
+        new KeyValueMapper[String, (String, JLong), KeyValue[String, JLong]] {
+          def apply(k: String, regionWithClicks: (String, JLong)) =
+            new KeyValue[String, JLong](regionWithClicks._1, regionWithClicks._2)
+        }
+      }
 
     // Compute the total per region by summing the individual click counts per region.
     val clicksPerRegion: KTableJ[String, JLong] = clicksByRegion
       .groupByKey(Grouped.`with`[String, JLong](Serdes.String, Serdes.JavaLong))
-      .reduce((v1: JLong, v2: JLong) => v1 + v2)
+      .reduce {
+        new Reducer[JLong] {
+          def apply(v1: JLong, v2: JLong) = v1 + v2
+        }
+      }
 
     // Write the (continuously updating) results to the output topic.
     clicksPerRegion.toStream.to(outputTopicJ, Produced.`with`(Serdes.String, Serdes.JavaLong))

--- a/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
+++ b/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
@@ -22,6 +22,8 @@ package org.apache.kafka.streams.scala
 import java.util.regex.Pattern
 
 import org.apache.kafka.streams.kstream.{
+  KeyValueMapper,
+  Reducer,
   Transformer,
   TransformerSupplier,
   ValueJoiner,
@@ -117,7 +119,11 @@ class TopologyTest extends JUnitSuite {
         }
       )
 
-      val grouped: KGroupedStreamJ[String, String] = splits.groupBy((_: String, v: String) => v)
+      val grouped: KGroupedStreamJ[String, String] = splits.groupBy {
+        new KeyValueMapper[String, String, String] {
+          def apply(k: String, v: String): String = v
+        }
+      }
 
       grouped.count()
 
@@ -175,15 +181,21 @@ class TopologyTest extends JUnitSuite {
 
       // Change the stream from <user> -> <region, clicks> to <region> -> <clicks>
       val clicksByRegion: KStreamJ[String, JLong] = userClicksJoinRegion
-        .map(
-          (_: String, regionWithClicks: (String, JLong)) =>
-            new KeyValue[String, JLong](regionWithClicks._1, regionWithClicks._2)
-        )
+        .map {
+          new KeyValueMapper[String, (String, JLong), KeyValue[String, JLong]] {
+            def apply(k: String, regionWithClicks: (String, JLong)) =
+              new KeyValue[String, JLong](regionWithClicks._1, regionWithClicks._2)
+          }
+        }
 
       // Compute the total per region by summing the individual click counts per region.
       clicksByRegion
         .groupByKey(Grouped.`with`[String, JLong])
-        .reduce((v1: JLong, v2: JLong) => v1 + v2)
+        .reduce {
+          new Reducer[JLong] {
+            def apply(v1: JLong, v2: JLong) = v1 + v2
+          }
+        }
 
       builder.build().describe()
     }
@@ -241,7 +253,11 @@ class TopologyTest extends JUnitSuite {
             }
         })
 
-      val grouped: KGroupedStreamJ[String, String] = lowered.groupBy((_: String, v: String) => v)
+      val grouped: KGroupedStreamJ[String, String] = lowered.groupBy {
+        new KeyValueMapper[String, String, String] {
+          def apply(k: String, v: String): String = v
+        }
+      }
 
       grouped.count()
 

--- a/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
+++ b/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
@@ -46,6 +46,7 @@ import _root_.scala.collection.JavaConverters._
 /**
  * Test suite that verifies that the topology built by the Java and Scala APIs match.
  */
+//noinspection ConvertExpressionToSAM due to 2.11 build
 class TopologyTest extends JUnitSuite {
 
   private val inputTopic = "input-topic"
@@ -193,7 +194,7 @@ class TopologyTest extends JUnitSuite {
         .groupByKey(Grouped.`with`[String, JLong])
         .reduce {
           new Reducer[JLong] {
-            def apply(v1: JLong, v2: JLong) = v1 + v2
+            def apply(v1: JLong, v2: JLong): JLong = v1 + v2
           }
         }
 
@@ -214,7 +215,6 @@ class TopologyTest extends JUnitSuite {
       val streamBuilder = new StreamsBuilder
       val textLines = streamBuilder.stream[String, String](inputTopic)
 
-      //noinspection ConvertExpressionToSAM due to 2.11 build
       val _: KTable[String, Long] =
         textLines
           .transform(new TransformerSupplier[String, String, KeyValue[String, String]] {

--- a/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
+++ b/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
@@ -147,6 +147,7 @@ class TopologyTest extends JUnitSuite {
 
       val userRegionsTable: KTable[String, String] = builder.table(userRegionsTopic)
 
+      // clicks per region
       userClicksStream
         .leftJoin(userRegionsTable)((clicks, region) => (if (region == null) "UNKNOWN" else region, clicks))
         .map((_, regionWithClicks) => regionWithClicks)

